text2digits/rules.py

```diff
@@ -0,0 +1,174 @@
+import enum
+from abc import ABC, abstractmethod
+from decimal import Decimal
+from typing import List, Union
+
+from text2digits.tokens_basic import WordType, Token, NoneToken
+from text2digits.tokens_rules import CombinedToken, ConcatenatedToken, RuleToken
+
+
+class Rule(ABC):
+    """
+    Rules are used to parse a sequence of tokens and to apply a rule on the detected tokens to retrieve a new number representation.
+    """
+
+    @abstractmethod
+    def match(self, tokens: List[Union[Token, RuleToken]]) -> int:
+        """
+        Analyses the tokens and tries to find a consecutive sequence of tokens which should be combined for the specified rule. The focus of this function lies on *which* tokens should be combined (instead of *how*).
+
+        :param tokens: List of tokens.
+        :return: Number of tokens which match the specified rule.
+        """
+        pass
+
+    @abstractmethod
+    def action(self, tokens: List[Union[Token, RuleToken]]) -> RuleToken:
+        """
+        Combines the tokens and replaces it with a new token (e.g. converted number). The focus of this function lies on *how* tokens should be combined (instead of *which*).
+
+        :param tokens: The tokens to be combined.
+        :return: The new token which replaces the input tokens.
+        """
+        pass
```

Comment:
empty on purpose? is it needed?

Reply:
Yes, this is the abstract class for all the rules. It enforces that every rule must have a `match` and an `action` method. Further, it allows to document the two methods only once instead for all rules ;-)